language/testing-infra/transactional-test-runner/src/tasks.rs

```diff
@@ -0,0 +1,319 @@
+                    start_line,
+                    command_lines_stop,
+                    help
+                )
+            }
+        };
+        let name = name_opt.unwrap();
+
+        let stop_line = if text.is_empty() {
+            command_lines_stop
+        } else {
+            text[text.len() - 1].0
+        };
+
+        // Keep fucking this up somehow
+        // let last_non_whitespace = text
+        //     .iter()
+        //     .rposition(|(_, l)| !WHITESPACE.is_match(l))
+        //     .unwrap_or(0);
+        // let initial_text = text
+        //     .into_iter()
+        //     .take_while(|(i, _)| *i < last_non_whitespace)
+        //     .map(|(_, l)| l);
+        let file_text_vec = (0..command_lines_stop)
+            .map(|_| String::new())
+            .chain(text.into_iter().map(|(_ln, l)| l))
+            .collect::<Vec<String>>();
+        let data = if file_text_vec.iter().all(|s| WHITESPACE.is_match(s)) {
+            None
+        } else {
+            let data = NamedTempFile::new()?;
+            data.reopen()?
+                .write_all(file_text_vec.join("\n").as_bytes())?;
+            Some(data)
+        };
+
+        tasks.push(TaskInput {
+            command,
+            name,
+            number,
+            start_line,
+            command_lines_stop,
+            stop_line,
+            data,
+        })
+    }
+    Ok(tasks)
+}
+
+impl<T: Debug + StructOpt> TaskInput<T> {
+    pub fn split<Subcommand: Debug + StructOpt>(
+        self,
+        split: impl FnOnce(T) -> Either<ProvidedCommand, Subcommand>,
+    ) -> Either<TaskInput<ProvidedCommand>, TaskInput<Subcommand>> {
+        let Self {
+            command,
+            name,
+            number,
+            start_line,
+            command_lines_stop,
+            stop_line,
+            data,
+        } = self;
+        match split(command) {
+            Either::Left(cmd) => Either::Left(TaskInput {
+                command: cmd,
+                name,
+                number,
+                start_line,
+                command_lines_stop,
+                stop_line,
+                data,
+            }),
+            Either::Right(subcmd) => Either::Right(TaskInput {
+                command: subcmd,
+                name,
+                number,
+                start_line,
+                command_lines_stop,
+                stop_line,
+                data,
+            }),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum SyntaxChoice {
+    Source,
+    IR,
+}
+
+#[derive(Debug, StructOpt)]
+pub enum ProvidedCommand {
+    #[structopt(name = "publish")]
+    Publish {
+        #[structopt(long = "gas-budget")]
+        gas_budget: Option<u64>,
+        #[structopt(long = "syntax")]
+        syntax: Option<SyntaxChoice>,
+        #[structopt(long = "address", parse(try_from_str = parse_account_address))]
+        address: Option<AccountAddress>,
```

Comment:
We may need to change this to a string to support named addresses, or make the address an associated type so that the adapter can decide what to do with it.

Reply:
Not sure if I follow

Reply:
Just wondering if we'll ever want to use named addresses here in the future. (Totally fine with it being a hex address for now.)

Reply:
Yup, we likely will need to